collector/collector.go

```diff
@@ -27,119 +27,146 @@ type CollectorOpts struct {
+		Log:                      c.log,
+		UID:                      c.opts.UID,
+		Location:                 c.opts.Location,
+		OutDir:                   c.opts.OutDir,
+		CheckNodeURI:             c.opts.CheckNodeURI,
+		ClickhouseDSN:            c.opts.ClickhouseDSN,
+		HTTPReceivers:            c.opts.Receivers,
+		ReceiversAllowedSources:  c.opts.ReceiversAllowedSources,
+		ReceiversAllowAllSources: c.opts.ReceiversAllowAllSources,
+		APIServer:                apiServer,
 	})
 
-	// If API server is running, add it as a TX receiver
-	if apiServer != nil {
-		processor.receivers = append(processor.receivers, apiServer)
-	}
-
 	// Start the transaction processor, which kicks off background goroutines
-	processor.Start()
+	c.processor.Start()
 
 	// Connect to regular nodes
-	for _, node := range opts.Nodes {
-		conn := NewNodeConnection(opts.Log, node, processor.txC)
+	for _, node := range c.opts.Nodes {
+		conn := NewNodeConnection(c.log, node, c.processor.txC)
 		conn.StartInBackground()
 	}
 
 	// Connect to Bloxroute
-	for _, auth := range opts.BloxrouteAuth {
+	for _, auth := range c.opts.BloxrouteAuth {
 		token, url := common.GetAuthTokenAndURL(auth)
 		startBloxrouteConnection(BlxNodeOpts{
-			TxC:        processor.txC,
-			Log:        opts.Log,
+			TxC:        c.processor.txC,
+			Log:        c.log,
 			AuthHeader: token,
 			URL:        url,
 		})
 	}
 
 	// Connect to Eden
-	for _, auth := range opts.EdenAuth {
+	for _, auth := range c.opts.EdenAuth {
 		token, url := common.GetAuthTokenAndURL(auth)
 		startEdenConnection(EdenNodeOpts{
-			TxC:        processor.txC,
-			Log:        opts.Log,
+			TxC:        c.processor.txC,
+			Log:        c.log,
 			AuthHeader: token,
 			URL:        url,
 		})
 	}
 
 	// Connect to Chainbound
-	for _, auth := range opts.ChainboundAuth {
+	for _, auth := range c.opts.ChainboundAuth {
 		token, url := common.GetAuthTokenAndURL(auth)
 		chainboundConn := NewChainboundNodeConnection(ChainboundNodeOpts{
-			TxC:    processor.txC,
-			Log:    opts.Log,
+			TxC:    c.processor.txC,
+			Log:    c.log,
 			APIKey: token,
 			URL:    url,
 		})
 		go chainboundConn.Start()
 	}
 }
+
+func (c *Collector) StartAPIServer() *api.Server {
+	if c.opts.APIListenAddr == "" {
+		return nil
+	}
+	apiServer := api.New(&api.HTTPServerConfig{
+		Log:        c.log,
+		ListenAddr: c.opts.APIListenAddr,
+	})
+	go apiServer.RunInBackground()
+	return apiServer
+}
+
+func (c *Collector) StartMetricsServer() {
+	if c.opts.MetricsListenAddr == "" {
+		return
+	}
+	mux := chi.NewRouter()
+
+	// Add regular routes
+	mux.HandleFunc("/metrics", func(w http.ResponseWriter, r *http.Request) {
+		metrics.WritePrometheus(w, true)
+	})
+	mux.HandleFunc("/livez", func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		_, _ = w.Write([]byte("ok"))
+	})
+	mux.HandleFunc("/readyz", func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		_, _ = w.Write([]byte("ok"))
```

Comment:
It would be ideal for readiness check to check if the startup have finished (in our case, all sources connected and data is flowing), but if it's too much refactor, that's good for now

Reply:
https://github.com/flashbots/mempool-dumpster/issues/77

Reply:
let's do it as separate pr